### ISSUE TITLE
save nvidia 3d Settings also

### ISSUE THREAD

**Issue by gretzer** [terrymacdonald/DisplayMagician#53]
**Is your feature request related to a problem? Please describe.**
I've got a Triple Screen + 1 (on top) simrig setup. I also use this machine (with nvidia GPU) for working/gaming at an usual desk. So i have to kneel everytime to plug one cable out and the other in. That's a problem this software definetly can not solve. :) Fortunately this little helper saves me the work of always rearranging nvidia surround and displays. Thank you!
One thing what is missing. My triple setup runs on 165Hz - so i put the max FPS limit in nvidia 3d settings to 162FPS. When I switch to single setup I set the max FPS to 98 FPS, since it's a 100Hz display.

**Describe the solution you'd like**
Not only save Display arrangement but also the 3d settings.

**Describe alternatives you've considered**
Doing it manually or doing it with a 3rd software is possible but not comfortable.

**Additional context**
Thanks again, I've searched months for this tool and it's a great experience!

**Comment by terrymacdonald** [comment]
I had a quick look at this the other day, and I think that when I solve #50, I'll solve this one too. It looks like those settings are part of the SetDisplayConfig and GetDisplayConfig NVAPI calls. The problem is that those two functions are the very two I can't get to work in my current testing :). 

I'll keep plugging away at this when I have some spare time, and see if magically I can unblock things. For this reason I'll mark it as a future enhancement rather than mark it for a particular release.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@gretzer I'm struggling to find where those settings you described are. Can you please take a screenshot of those settings and describe where you set them?

Can you also fill out the following list to describe your hardware and software setup please:
 - Windows Version: [e.g. Win10]
 - DisplayMagician Version [e.g. 1.1.0]
 - Video Card: [e.g. Asus NVIDIA GTX1070]
 - Number and make/model of monitors: [e.g. 2xDell UH2718H, 1x LG G27U17D]
 - Date of last video driver update: [e.g. 2021-02-15]
 - Date of last windows update: [e.g. 2021-02-15]

Thanks
Terry

**Comment by terrymacdonald** [comment]
Further digging through the NVAPI points me to NvAPI_D3D_GetSleepStatus and NvAPI_D3D_SetSleepMode... I'll leave these here as a future reminder.

// FUNCTION NAME: NvAPI_D3D_GetSleepStatus
//
//!   DESCRIPTION: This function can be used to get the latest sleep status.
//!   bLowLatencyMode indicates whether low latency mode is currently
//!   enabled in the driver.
//!   Note that it may not always reflect the previously requested sleep mode,
//!   as the feature may not be available on the platform, or the setting has
//!   been overridden by the control panel, for example.

// FUNCTION NAME: NvAPI_D3D_SetSleepMode
//
//!   DESCRIPTION: This function can be used to update sleep mode dynamically.
//!   The settings are not dependent to each other, meaning low latency mode
//!   can be enabled/disabled regardless of whether minimum interval is set or
//!   not. The former is to intelligently lower latency without impacting frame
//!   rate. The later is to limit frame rate (e.g. minimumIntervalUs = 10000
//!   limits frame rate to 100 FPS). They work well separately and/or together.
//!   Note that minimumIntervalUs usage is not limited to lowering latency, so
//!   feel free to use it to limit frame rate for menu, cut scenes, etc.
//!   Note that low latency mode can be enabled, and/or minimum interval can
//!   be set, even without using NvAPI_D3D_Sleep(). However, without it, the
//!   sleep to achieve these features would happen at a less optimal point,
//!   resulting in higher overall latency.
//!   The bLowLatencyBoost will request the GPU run at max clocks even in 
//!   scenarios where it is idle most of the frame and would normally try
//!   to save power.  This can decrease latency in CPU-limited scenarios.
//!   While this function can be called as often as needed, it is not
//!   necessary nor recommended to call this too frequently (e.g. every frame),
//!   as the settings persist for the target device.

**Comment by gretzer** [comment]
Hi Terry,

thanks for your reply. I'm currently not sitting at my rig. I'm trying to answer your questions from my memory.

- Windows 10
- DisplayMagician 2.0.1
- RTX 3080 Founders Edition
- 1x LC-Power LC-M34-UWQHD-100-C; 3x LG 32GN600-B; 1x Benq GL2450-T
- Last video driver update: End of October 21
- Last windows update: Beginning of November 21

Here you can see the 3d settings and the max. FPS usage:
https://www.howtogeek.com/509097/how-to-set-a-maximum-frame-rate-in-nvidias-drivers/

Also found a new workaround by setting program specific 3d settings (ac2.exe), this is possible in the nvidia control panel. So I use global 98FPS (when using the 34" single screen) and the 162FPS are set when ACC is started. But probably out there are some similare use cases where this would not fulfill their need to save 3d settings with monitor setup.

greets

**Comment by terrymacdonald** [comment]
Hi @gretzer,

I've just finished DisplayMagician v2.3.0.35 which is hopefully the next big release of DisplayMagician. It has a lot of improvements and bug fixes to the NVIDIALibrary, and it's possible that it will solve the issues that you've been experiencing. NVIDIALibrary can now handle multiple physical adapters in a logical group, dvi-d and earlier connections, USB connected displays, adaptive sync, image scaling, cloned displays and more. 

I *think* it covers your FPS usage settings. Can you please test that it does for me?

[DisplayMagicianSetup-v2.3.0.35.zip](https://github.com/terrymacdonald/DisplayMagician/files/8457813/DisplayMagicianSetup-v2.3.0.35.zip)

Can you please download and install DisplayMagician v2.3.0.35, and then do the following two steps:

- Delete your old Display Profiles 
- Create new Display Profiles 
- Update your Game Shortcuts to use the new Display Profiles

If you have any issues, please create a Support ZIP File (main screen > Settings > Create Support ZIP File) and upload it here through your web browser. I'll then be able to use that for any bug fixing.
 
Thanks
Terry

**Comment by gretzer** [comment]
Hi Terry,

thanks for your effort. After installation of the new version it unfortunately does not start anymore.

![image](https://user-images.githubusercontent.com/84306602/162679147-d8b3f5c4-15b1-4a3b-ad87-cace151427c9.png)

any hint?

**Comment by terrymacdonald** [comment]
Can you please send me your DisplayMagician.log files (should be 4 of them)
in your C:\Users\<Username>\AppData\Local\Displaymagician\Logs folder? That
should tell me where the error lies, and give me an idea of where to fix.

Oh, also please upload through the Display profiles.json in case that helps.

Is that screenshot what you got straight after the upgrade?

Thanks
Terry

**Comment by gretzer** [comment]
Hey Terry,

of course. Here!

[Logs.zip](https://github.com/terrymacdonald/DisplayMagician/files/8461744/Logs.zip)
[Profiles.zip](https://github.com/terrymacdonald/DisplayMagician/files/8461746/Profiles.zip)


greets

**Comment by terrymacdonald** [comment]
Gah! The issue is my botched Display Profile upgrade logic :(.  Now fixed.

Please do the following:

- Delete `C:\Users\<Username>\AppData\Local\Displaymagician\Profiles\DisplayProfiles_2.3.json`
- Install this new DisplayMagician version:
[DisplayMagicianSetup-v2.3.0.36.zip](https://github.com/terrymacdonald/DisplayMagician/files/8467396/DisplayMagicianSetup-v2.3.0.36.zip)
- Run DisplayMagician
- Create your Display Profiles again the way you want them
- Update your Game Shortcuts to use the new Display Profiles
- Have fun with your games!

Hopefully it should work this time.  Please log a new issue at https://github.com/terrymacdonald/DisplayMagician/issues/new/choose if you experience any new bugs that you've not seen before, and I'll get right on fixing them.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@gretzer I've found the cause of the error you were seeing, and I've _partially_ fixed it. I still have a bit more work to go on it, but should have a new build for you to try in the next few days.

**Comment by terrymacdonald** [comment]
@gretzer New build time! I *think* I've fixed the issues you've experienced. 

I've completely changed the way that display profile layout images have are created, and it's a lot simpler, so should avoid some of the issues that you would have experienced. I've also found some bugs with the rebuild of the ShortcutForm and ShortcutItem that I did in v2.3 and fixed those, so that should stop some of the issues that occurred when creating new Game Shortcuts. 

[DisplayMagicianSetup-v2.3.0.50.zip](https://github.com/terrymacdonald/DisplayMagician/files/8488041/DisplayMagicianSetup-v2.3.0.50.zip)

So can you please download and install DisplayMagician v2.3.0.50 from here and give it a test? If you find any issues please create a  Support ZIP File and upload to this issue.

Thanks!
Terry

**Comment by terrymacdonald** [comment]
@gretzer Another build with a couple of additional fixes that other users have reported.

The exception error when showing/saving a new profile should be fixed now. As should the detecting of the main game shutting down. You can test the new version works by downloading DisplayMagician v2.1.3.12 from here:
[DisplayMagicianSetup-v2.3.1.12.zip](https://github.com/terrymacdonald/DisplayMagician/files/8499351/DisplayMagicianSetup-v2.3.1.12.zip)

There is only one more small bug that I'm aware of that I'm trying to fix but it only seems to happen with some specific start programs and only when run from a desktop shortcut. Hoping to fix that within the next few days.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@gretzer I have worked through many issues over the last week, so I recommend that you do some testing with this version:
[DisplayMagicianSetup-v2.3.1.50.zip](https://github.com/terrymacdonald/DisplayMagician/files/8512144/DisplayMagicianSetup-v2.3.1.50.zip)

I **strongly** recommend that  you delete your old display profiles and recreate them again just to be 100% sure you're running the latest versions of the file, and then updated your game shortcuts to use those new profiles.

This version fixes all the bugs that I know about, and based on my testing is just about ready to release to the general public. I'd really appreciate it if you could do some testing (but only if you have some spare time) so that I will know if it really does fix all the issues.

Thanks
Terry

**Comment by gretzer** [comment]
Hello Terry!

First of all, sorry for my late response. the last days were busy...

I tested v2.3.1.50. So far the software runs without crashes. But the maximum FPS in the nvidia control panel is not changing when I change the profile. :(

greets 
gretzer

**Comment by terrymacdonald** [comment]
Hi @gretzer,

I've found an issue with the colour of the screen in a surround profile, and I've corrected it in this version. It should fix all the profile display errors (but only will do so when you create a new display profile and have edited a shortcut using that profile). 

[DisplayMagicianSetup-v2.3.1.56.zip](https://github.com/terrymacdonald/DisplayMagician/files/8516635/DisplayMagicianSetup-v2.3.1.56.zip)

That's really disappointing it doesn't track and store the maximum FPS setting. Can you please do a new Support ZIP File so that I can see what settings are in your display config? Can you do me a couple of screenshots of the NVIDIA Control Panel settings you changed (one for each display profile so I can see what you set them to)? 

Is the specific setting the 'Max Frame Rate' under the 'Manage 3D Settings' window in the NVIDIA Control Panel? What values did you set it to?

Thanks
Terry

**Comment by terrymacdonald** [comment]
@gretzer, I finally found where the settings are stored! They are in the Driver Settings section, which is separate from the other functions I've been searching through: https://docs.nvidia.com/gameworks/content/gameworkslibrary/coresdk/nvapi/group__drsapi.html#func-members

The bad news is that there is quite a lot of work to build the structures required to process and store these settings, write the interactions with the driver functions, figure out which ones are the ones I should use, and actually write the storage code, start on the application code, figure out how to apply the settings reliably without affecting the other settings, then writing the final code. **I won't be able to get the logic into the v2.3.1 release.**

Instead, I'm preparing for the future by adding a space for them in the config file. At present DM writes an empty object, but it will mean that I'll be able to add this in a future release without breaking everyone's display profiles and forcing them to recreate them.

So, I'll have to bump this to a future release of DisplayMagician. It won't be built for a few versions I think just based on the work required to add it.

Sorry for the bad news.
Thanks
Terry

**Comment by terrymacdonald** [comment]
@gretzer Turns our @ceramod needs this same feature to turn on the high performance GPU in the Optimus. I'm planning on adding this in v2.4.0 but there will be a bit of work to get this done. I'm expecting it to take a few months before I can get through the amount of work it needs, but will update you to progress as I have been doing so far.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@gretzer time for an update. I now have it 'mostly' working within my test library of NVIDIAInfo, but I still have some more work to do to iron out a few bugs in the NVIDIALibrary. I have to give a massive shout out to Soroush Falahati (the author of NvAPIWrapper.net and HeliosDisplayManagement) as I wouldn't have been able to get the DRS settings functionality to work with his awesome code to use as a base. That man is a legend.

So, I'm getting closer to having a test version of NVIDIAInfo for you to try again with. When I've fixed the last few bugs we'll do some testing and see if any more bugs shake out. Then when it's working I then need to integrate the updated NVIDIALibrary into DisplayMagician.

We're on track for this being in DisplayMagician v2.4.0.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@gretzer So I now have a new version of NVIDIAInfo to test that includes DRS Settings. It now stores them in .cfg files and loads and applies them. Can you please perform the following instructions, and then that will give me the information I need to be sure that the new DRS settings features works.

Download NVIDIAInfo from here, and extract it to a folder:
[NVIDIAInfo-v1.7.4.zip](https://github.com/terrymacdonald/DisplayMagician/files/8692633/NVIDIAInfo-v1.7.4.zip)

Open a PowerShell window, and change to the same folder that you extracted NVIDIAInfo into.

**IMPORTANT - These instructions are likely to fail, so I need you to save/rename the NVIDIAInfo.log file as shown below. Each run of NVIDIAInfo overwrites the previous NVIDIAInfo.log, so I need you to rename it each time so it doesn't get overwritten. I need to see the log file entries to know what is broken.**

Please complete the following instructions in this specific order:

Create your surround display layout with some of your screens (surround):
- Save the first surround display as `NVIDIAInfo save surround.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-surround-save.log"`
- Check if the first windows display config is equal with `NVIDIAInfo equal surround.cfg` (we want it to be equal)

Next create a second windows display layout (regular):
- Create a regular display layout with different screens (turn some of the original screens off, and move other ones around, and change the primary monitor)
- Save the second windows display as `NVIDIAInfo save regular.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-regular-save.log"`
- Check if the second windows display config is equal with `NVIDIAInfo equal regular.cfg` (we want it to be equal)

Next swap back to the surround display layout:
- Load the first windows display as `NVIDIAInfo load surround.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-surround-reload.log"`
- Check if the first windows display config is equal with `NVIDIAInfo equal surround.cfg` (we want it to be equal)
- Save a copy of the current display config for me to compare: `NVIDIAInfo save surround-reload.cfg`
- Make sure the second windows display config is NOT equal with `NVIDIAInfo equal regular.cfg` (we DO NOT want it to be equal)

Next swap back to the regular display layout:
- Load the second windows display as `NVIDIAInfo load regular.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-regular-reload.log"`
- Check if the second windows display config is equal with `NVIDIAInfo equal regular.cfg` (we want it to be equal)
- Save a copy of the current display config for me to compare: `NVIDIAInfo save regular-reload.cfg`
- Make sure the first windows display config is NOT equal with `NVIDIAInfo equal surround.cfg` (we DO NOT want it to be equal)

Once that series of tests are done, please include the following files in a ZIP file, and upload the ZIP to this github issue through your web browser (click and drag the file from your File explorer into Github and it will upload):
- surround.cfg
- surround-reload.cfg
- NVIDIAInfo-surround-save.log
- NVIDIAInfo-surround-reload.log
- regular.cfg
- regular-reload.cfg
- NVIDIAInfo-regular-save.log
- NVIDIAInfo-regular-reload.log

I know this is a looooooong list of instructions, but it really is key to do them in order so that if there are errors then I can piece together what happened and find a fix.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@gretzer How did your testing go? You should have the DRS Settings (3d Settings in the NVIDIA Control Panel) viewable in the cfg files after you've saved a couple, e.g.:

````
           }
            ....
             "Name": "Toggle the VRR global feature",
            "SettingId": 278196567,
            "SettingType": 0,
            "SettingLocation": 0,
            "IsCurrentPredefined": false,
            "IsPredefinedValid": false,
            "PredefinedValue": null,
            "CurrentValue": 1
          },
````

At this point I will mark this issue as closed, as the DRS is now part of the main NVIDIALibrary in NVIDIAInfo. This will then be released as part of DisplayMagician v2.4.0 in a month or so.

If it turns out this is still broken on your system and your NVIDIAInfo changes don't work, then we can reopen this issue and continue to work on it.

Thanks
Terry
